### PR TITLE
fix: parse report.jsonl without creating large string

### PR DESCRIPTION
**Pull request by yury-s**
Reference https://github.com/microsoft/playwright/issues/28362

### CODE REVIEW

File: packages/playwright/src/reporters/merge.ts [hunk 1]
```diff
@@ -81,19 +81,35 @@ const commonEvents = new Set(commonEventNames);
...
-  return reportJsonl.toString().split('\n')
+  return splitBufferLines(reportJsonl)
+      .map(line => line.toString('utf8'))
       .filter(line => line.length)
       .map(line => JSON.parse(line) as JsonEvent)
       .filter(event => !commonEvents.has(event.method));
 }
 
+function splitBufferLines(buffer: Buffer) {
+  const lines = [];
```
[the remark this answers is not in the document]
Comment (yury-s) [reply packages/playwright/src/reporters/merge.ts:100]: created buffer 197 MB
splitNative: 47ms
splitBufferLines: 723ms


```js
const lines = [];
for (let i = 0; i < 100_000; i++) {
  lines.push(`a${i}_`.repeat(300));
}

const string = lines.join('\n');

const buffer = Buffer.from(string);

for (let i = 0; i < buffer.length; i++) {
  if (buffer[i] === 1) {
    console.log('found at', i);
  }
}

console.log('created buffer', buffer.length >> 20, 'MB');
withTimer('splitNative', () => splitNative(buffer));
withTimer('splitBufferLines', () => splitBufferLines(buffer));

function withTimer(name, fn) {
  const start = Date.now();
  const result = fn();
  const end = Date.now();
  console.log(`${name}: ${end - start}ms`);
  return result;
}

function splitNative(buffer) {
  return buffer.toString().split('\n');
}

function splitBufferLines(buffer) {
  const lines = [];
  let start = 0;
  for (let i = 0; i < buffer.length; i++) {
    if (buffer[i] === 0x0A) { // 0x0A is the byte for '\n'
      lines.push(buffer.slice(start, i));
      start = i + 1;
    }
  }
  if (start < buffer.length)
    lines.push(buffer.slice(start));
  return lines.map(b => b.toString());
}

```

File: packages/playwright/src/reporters/merge.ts [hunk 2]
```diff
@@ -81,19 +81,35 @@ const commonEvents = new Set(commonEventNames);
...
+  let start = 0;
+  for (let i = 0; i < buffer.length; i++) {
```
[the remark this answers is not in the document]
Comment (yury-s) [reply packages/playwright/src/reporters/merge.ts]: Done. It filters out last "empty" buffer, but that's okay for our use case.

File: packages/playwright/src/reporters/merge.ts [hunk 3]
```diff
@@ -81,19 +81,37 @@ const commonEvents = new Set(commonEventNames);
...
+      .map(line => line.toString('utf8'))
       .filter(line => line.length)
       .map(line => JSON.parse(line) as JsonEvent)
       .filter(event => !commonEvents.has(event.method));
 }
 
+function splitBufferLines(buffer: Buffer) {
+  const lines = [];
+  let start = 0;
+  while (start < buffer.length) {
```
Comment (mxschmitt) [inline packages/playwright/src/reporters/merge.ts:102]: Maybe we should do something like this instead?

```ts
    const rl = readline.createInterface({
        input: fileStream,
        crlfDelay: Infinity
    });

    for await (const line of rl) {
        ...
    }
```

but probably not a huge difference if at the end the lines are anyways in the memory.

Comment (yury-s) [reply packages/playwright/src/reporters/merge.ts:102]: Yeah, the buffer is in memory anyway, so I'd keep it simple for now.

**Review by dgozman — APPROVED** [review]

**Pull request merged**